# Visual Data Analysis of Fraudulent Transactions

Your CFO has also requested detailed trends data on specific card holders. Use the starter notebook to query your database and generate visualizations that supply the requested information as follows, then add your visualizations and observations to your markdown report.

In [31]:
# Initial imports
import pandas as pd
import calendar
import hvplot.pandas
from sqlalchemy import create_engine

In [32]:
# Create a connection to the database
engine = create_engine("postgresql://postgres:abc@localhost:5432/fraud_detection")


## Data Analysis Question 1

The two most important customers of the firm may have been hacked. Verify if there are any fraudulent transactions in their history. For privacy reasons, you only know that their cardholder IDs are 2 and 18.

* Using hvPlot, create a line plot representing the time series of transactions over the course of the year for each cardholder separately. 

* Next, to better compare their patterns, create a single line plot that containins both card holders' trend data.  

* What difference do you observe between the consumption patterns? Does the difference suggest a fraudulent transaction? Explain your rationale in the markdown report.

In [33]:
# loading data for card holder 2 and 18 from the database
# Write the query
query = 'SELECT * FROM transaction as a INNER JOIN credit_card as b ON(a.card = b.card) WHERE cardholder_id = 2 or cardholder_id = 18;'

cardholder_df = pd.read_sql(query, engine)

# Create a DataFrame from the query result. HINT: Use pd.read_sql(query, engine)
cardholder_df


id                date   amount                 card  id_merchant  \
0     567 2018-01-01 23:15:10    $2.95        4498002758300           64   
1    2077 2018-01-05 07:19:27    $1.36      344119623920892           30   
2    2439 2018-01-06 02:16:41    $1.33  4866761290278198714          127   
3    1867 2018-01-06 05:13:20   $10.82  4866761290278198714           70   
4    3457 2018-01-07 01:10:54  $175.00      344119623920892           12   
..    ...                 ...      ...                  ...          ...   
227  1994 2018-12-27 18:46:57    $1.70      344119623920892           55   
228   114 2018-12-28 08:45:26    $3.46        4498002758300           82   
229  1228 2018-12-28 09:00:45   $12.88      344119623920892           60   
230   962 2018-12-28 15:30:55   $11.03         675911140852            2   
231  1218 2018-12-29 08:11:55   $12.25        4498002758300            6   

                    card  cardholder_id  
0          4498002758300             18  
1        344119623920892             18  
2    4866761290278198714              2  
3    4866761290278198714              2  
4        344119623920892             18  
..                   ...            ...  
227      344119623920892             18  
228        4498002758300             18  
229      344119623920892             18  
230         675911140852              2  
231        4498002758300             18  

[232 rows x 7 columns]

In [34]:
# Clean up data
card2 = cardholder_df.loc[cardholder_df["cardholder_id"] == 2]
card2= card2.drop(columns= ["card","id_merchant", "cardholder_id", "id"])
card2["amount"] = card2["amount"].replace({'\$':''}, regex = True)
card2["amount"] = card2["amount"].apply(pd.to_numeric, downcast='float', errors='coerce')

# Plot for Cardholder 2
plot2 = card2.hvplot.line(
    label="Time Series of Transactions over a year for Cardholder 2",
    x = "date",
    y = "amount",
    xlabel = "Time",
    ylabel = "Amount ($)",
    ylim = (.50, 25),
    hover_line_color = "green")

plot2


:Curve   [date]   (amount)

In [35]:
# Plot for cardholder 18
card18 = cardholder_df.loc[cardholder_df["cardholder_id"] == 18]
card18= card18.drop(columns= ["card","id_merchant", "cardholder_id", "id"])
card18["amount"] = card18["amount"].replace({'\$':''}, regex = True)
card18["amount"] = card18["amount"].apply(pd.to_numeric, downcast='float', errors='coerce')

# Plot for cardholder 18
plot18 = card18.hvplot.line(
    label="Time Series of Transactions over a year for Cardholder 18",
    x = "date",
    y = "amount",
    xlabel = "Time",
    ylabel = "Amount ($)",
    ylim = (0.50, 1000),
    hover_line_color = "green")

plot18

:Curve   [date]   (amount)

In [36]:
# Combined plot for card holders 2 and 18
plot18 * plot2

:Overlay
   .Curve.Time_Series_of_Transactions_over_a_year_for_Cardholder_18 :Curve   [date]   (amount)
   .Curve.Time_Series_of_Transactions_over_a_year_for_Cardholder_2  :Curve   [date]   (amount)

What difference do you observe between the consumption patterns? Does the difference suggest a fraudulent transaction? Explain your rationale.
Cardholder 2 has lower spending amounts and it is very consistent, while Cardholder 18 has spike in their spending amount in Jan, July, August and September. It would be good to validate outlier transaction with cardholder 18.


## Data Analysis Question 2

The CEO of the biggest customer of the firm suspects that someone has used her corporate credit card without authorization in the first quarter of 2018 to pay quite expensive restaurant bills. Again, for privacy reasons, you know only that the cardholder ID in question is 25.

* Using hvPlot, create a box plot, representing the expenditure data from January 2018 to June 2018 for cardholder ID 25.

* Are there any outliers for cardholder ID 25? How many outliers are there per month?

* Do you notice any anomalies? Describe your observations and conclusions in your markdown report.

In [37]:
# loading data of daily transactions from jan to jun 2018 for card holder 25
# Write the query
query = "SELECT * FROM transaction as a LEFT JOIN credit_card as b ON(a.card = b.card) WHERE date <'2018-7-01'and cardholder_id = 25;"

cardholder25_df = pd.read_sql(query, engine)

# Create a DataFrame from the query result. HINT: Use pd.read_sql(query, engine)
cardholder25_df

id                date     amount             card  id_merchant  \
0   2083 2018-01-02 02:06:21      $1.46    4319653513507           93   
1   1552 2018-01-05 06:26:45     $10.74  372414832802279           86   
2   2108 2018-01-07 14:57:23      $2.93    4319653513507          137   
3    754 2018-01-10 00:25:40      $1.39  372414832802279           50   
4   3023 2018-01-14 05:02:22     $17.84  372414832802279           52   
..   ...                 ...        ...              ...          ...   
63  2582 2018-06-22 06:16:50  $1,813.00    4319653513507           40   
64  3218 2018-06-23 22:36:00     $16.61    4319653513507          144   
65  1523 2018-06-25 09:17:05     $11.53  372414832802279           77   
66  2264 2018-06-27 14:33:06      $5.24  372414832802279           26   
67   671 2018-06-30 03:05:55      $2.27  372414832802279           87   

               card  cardholder_id  
0     4319653513507             25  
1   372414832802279             25  
2     4319653513507             25  
3   372414832802279             25  
4   372414832802279             25  
..              ...            ...  
63    4319653513507             25  
64    4319653513507             25  
65  372414832802279             25  
66  372414832802279             25  
67  372414832802279             25  

[68 rows x 7 columns]

In [38]:
# loop to change the numeric month to month names
import datetime
# add new month column
cardholder25_df["month"] = ""

# initialize variable
counter = 0

# loop through, change the numeric month to month names and store in month column
for x in cardholder25_df["date"]:
    cardholder25_df.loc[counter,"month"] = str(x.month_name())
    counter+=1

cardholder25_df.tail()

id                date     amount             card  id_merchant  \
63  2582 2018-06-22 06:16:50  $1,813.00    4319653513507           40   
64  3218 2018-06-23 22:36:00     $16.61    4319653513507          144   
65  1523 2018-06-25 09:17:05     $11.53  372414832802279           77   
66  2264 2018-06-27 14:33:06      $5.24  372414832802279           26   
67   671 2018-06-30 03:05:55      $2.27  372414832802279           87   

               card  cardholder_id month  
63    4319653513507             25  June  
64    4319653513507             25  June  
65  372414832802279             25  June  
66  372414832802279             25  June  
67  372414832802279             25  June

In [39]:
#Clean up data

cardholder25_df= cardholder25_df.drop(columns= ["card","id_merchant", "cardholder_id", "id", "date"])
cardholder25_df["amount"] = cardholder25_df["amount"].replace({'\$':''}, regex = True)
cardholder25_df["amount"] = cardholder25_df["amount"].replace({'\,':''}, regex = True)
cardholder25_df["amount"] = cardholder25_df["amount"].apply(pd.to_numeric, downcast='float')
cardholder25_df

c:\Users\Max Work\anaconda3\envs\dev\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
c:\Users\Max Work\anaconda3\envs\dev\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
c:\Users\Max Work\anaconda3\envs\dev\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = valu

amount    month
0      1.460000  January
1     10.740000  January
2      2.930000  January
3      1.390000  January
4     17.840000  January
..          ...      ...
63  1813.000000     June
64    16.610001     June
65    11.530000     June
66     5.240000     June
67     2.270000     June

[68 rows x 2 columns]

In [40]:
help(cardholder25_df.hvplot.box)

Help on method box in module hvplot.plotting.core:

box(y=None, by=None, *, box_alpha, box_cmap, box_color, box_fill_alpha, box_fill_color, box_hover_alpha, box_hover_color, box_hover_fill_alpha, box_hover_fill_color, box_hover_line_alpha, box_hover_line_color, box_line_alpha, box_line_cap, box_line_color, box_line_dash, box_line_join, box_line_width, box_muted, box_muted_alpha, box_muted_color, box_muted_fill_alpha, box_muted_fill_color, box_muted_line_alpha, box_muted_line_color, box_nonselection_alpha, box_nonselection_color, box_nonselection_fill_alpha, box_nonselection_fill_color, box_nonselection_line_alpha, box_nonselection_line_color, box_selection_alpha, box_selection_color, box_selection_fill_alpha, box_selection_fill_color, box_selection_line_alpha, box_selection_line_color, box_visible, box_width, cmap, outlier_alpha, outlier_color, outlier_fill_alpha, outlier_fill_color, outlier_hover_alpha, outlier_hover_color, outlier_hover_fill_alpha, outlier_hover_fill_color, outlier_h

In [44]:
# Creating the six box plots using hvPlot
cardholder25_df.hvplot.box(
    y = 'amount',
    by = 'month',
    color = "forestgreen",
    title = "Transaction from Jan to Jun 2018 for cardholder 25 ",
    height=400, width=400, legend=False,figsize=(30,15)
)


:BoxWhisker   [month]   (amount)

Are there any outliers for cardholder ID 25? How many outliers are there per month?
On average, there are 1.5 outliers per month. June and April have 3 outliers.
Do you notice any anomalies? Describe your observations and conclusions.
June and April have 3 outliers and high amount of spending. 
